### ISSUE TITLE
RNN to discover Luhn Checksum

### ISSUE THREAD

**Issue by bkj** [fchollet/keras#914]
For fun and practice, I'm trying to construct a network that can reproduce the behavior of the Luhn Checksum (https://en.wikipedia.org/wiki/Luhn_algorithm), which is an error-correcting code for validating credit card numbers.  I was wondering if anyone could suggest an architecture that is flexible enough to model the behavior.

Here's how I'm generating the training data:

```

from pprint import pprint
from random import shuffle
from faker import Factory
fake = Factory.create()


# Run the Luhn Checksum on a credit card number
def cardLuhnChecksumIsValid(card_number):
    """ checks to make sure that the card passes a luhn mod-10 checksum """
    
    sum = 0
    num_digits = len(card_number)
    oddeven = num_digits & 1
    
    for count in range(0, num_digits):
        digit = int(card_number[count])
        
        if not (( count & 1 ) ^ oddeven ):
            digit = digit * 2
        if digit > 9:
            digit = digit - 9
            
        sum = sum + digit
        
    return ( (sum % 10) == 0 )


# Change one digit of the credit card number by 1
# This is guaranteed to corrupt a valid credit card number
def alter(x):
    return str(int(x) + 10 ** np.random.choice(range(8)))

K = 10000

X = [str(fake.credit_card_number()) for i in range(K)] + [alter(fake.credit_card_number()) for i in range(K)]
shuffle(X)
Y = map(cardLuhnChecksumIsValid, X)

pprint(X[0:5])
'''
['3112871329791258',
 '3158896578325611',
 '4392681270152917',
 '6011989628511339',
 '869986969145200']
 '''

pprint(Y[0:5])
''' [False, True, False, True, False] '''

```

Any advice would be much appreciated -- I'm trying to get a feel for what different architectures are designed for.  I had tried breaking the strings into lists of ints, padding all to length 16, then using

```
model = Sequential()
model.add(Embedding(10, 128, input_length=16))
model.add(LSTM(128))
model.add(Dense(1))
model.add(Activation('sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', class_mode="binary")
```

but this didn't work particularly well.

Thanks!
Ben

**Comment by bkj** [comment]
Also, interestingly, I think, if you comment out the 

```
...
        if digit > 9:
            digit = digit - 9
...
```

from the Luhn Checksum, the results jump up to > 95% accurate in just a few small training rounds.  If you leave that condition in, the network is no better than chance.  Intuitively, this seems like the kind of thing that would be addressed by adding another layer to the network (multiplication by 2 in the first layer, subtraction of 9 in the second layer if necessary), but that's really just a guess.

~ Ben

**Comment by fchollet** [comment]
You can check out the addition example, it's somewhat related. https://github.com/fchollet/keras/blob/master/examples/addition_rnn.py

**Comment by bkj** [comment]
Thanks for the suggestion.  I took a look at that and I see that it's operating on "sequence-to-sequence" data, while the Luhn checksum is a sequence classification problem.  If I modified the example to just output a single value, I think it would look like this:

```
model = Sequential()
model.add(RNN(HIDDEN_SIZE, input_shape=(None, len(chars))))
# model.add(RepeatVector(DIGITS + 1)) -- don't need this since we're not outputting sequence

for _ in xrange(LAYERS):
    model.add(RNN(HIDDEN_SIZE, return_sequences=True))

model.add(TimeDistributedDense(1)) # 1D for binary classification
model.add(Activation('sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam')
```

Is that correct, you think?  The differences between this and the archictecture I tried above that didn't work are using a time-distributed-dense layer instead of a dense layer, and not using an embedding -- possibly those changes plus adding multiple layers will give enough flexibility to learn the checksum?
